test/e2e/storage/pd.go

```diff
@@ -59,355 +60,255 @@ var _ = SIGDescribe("Pod Disks", func() {
-
-		framework.ExpectNoError(f.WaitForPodRunningSlow(host1ROPod.Name))
-
-		By("deleting host0ROPod")
-		framework.ExpectNoError(podClient.Delete(host0ROPod.Name, metav1.NewDeleteOptions(0)), "Failed to delete host0ROPod")
-
-		By("deleting host1ROPod")
-		framework.ExpectNoError(podClient.Delete(host1ROPod.Name, metav1.NewDeleteOptions(0)), "Failed to delete host1ROPod")
-
-		By("Test completed successfully, waiting for PD to safely detach")
-		waitForPDDetach(diskName, host0Name)
-		waitForPDDetach(diskName, host1Name)
-	})
-
-	It("Should schedule a pod w/ a readonly PD on two hosts, then remove both gracefully. [Slow]", func() {
-		framework.SkipUnlessProviderIs("gce", "gke")
-
-		By("creating PD")
-		diskName, err := framework.CreatePDWithRetry()
-		framework.ExpectNoError(err, "Error creating PD")
-
-		rwPod := testPDPod([]string{diskName}, host0Name, false /* readOnly */, 1 /* numContainers */)
-		host0ROPod := testPDPod([]string{diskName}, host0Name, true /* readOnly */, 1 /* numContainers */)
-		host1ROPod := testPDPod([]string{diskName}, host1Name, true /* readOnly */, 1 /* numContainers */)
-
-		defer func() {
-			By("cleaning up PD-RO test environment")
-			// Teardown pods, PD. Ignore errors.
-			// Teardown should do nothing unless test failed.
-			podClient.Delete(rwPod.Name, &metav1.DeleteOptions{})
-			podClient.Delete(host0ROPod.Name, &metav1.DeleteOptions{})
-			podClient.Delete(host1ROPod.Name, &metav1.DeleteOptions{})
-			detachAndDeletePDs(diskName, []types.NodeName{host0Name, host1Name})
-		}()
-
-		By("submitting rwPod to ensure PD is formatted")
-		_, err = podClient.Create(rwPod)
-		framework.ExpectNoError(err, "Failed to create rwPod")
-		framework.ExpectNoError(f.WaitForPodRunningSlow(rwPod.Name))
-		// Delete pod with default grace period 30s
-		framework.ExpectNoError(podClient.Delete(rwPod.Name, &metav1.DeleteOptions{}), "Failed to delete host0Pod")
-		framework.ExpectNoError(waitForPDDetach(diskName, host0Name))
-
-		By("submitting host0ROPod to kubernetes")
-		_, err = podClient.Create(host0ROPod)
-		framework.ExpectNoError(err, "Failed to create host0ROPod")
-
-		By("submitting host1ROPod to kubernetes")
-		_, err = podClient.Create(host1ROPod)
-		framework.ExpectNoError(err, "Failed to create host1ROPod")
-
-		framework.ExpectNoError(f.WaitForPodRunningSlow(host0ROPod.Name))
-
-		framework.ExpectNoError(f.WaitForPodRunningSlow(host1ROPod.Name))
-
-		By("deleting host0ROPod")
-		framework.ExpectNoError(podClient.Delete(host0ROPod.Name, &metav1.DeleteOptions{}), "Failed to delete host0ROPod")
-
-		By("deleting host1ROPod")
-		framework.ExpectNoError(podClient.Delete(host1ROPod.Name, &metav1.DeleteOptions{}), "Failed to delete host1ROPod")
+	Context("schedule a pod w/ a RW PD, delete pod, schedule it on another host, verify PD contents [Slow]", func() {
+		type testT struct {
+			descr     string                // It description
+			deleteOpt *metav1.DeleteOptions // pod delete option
```

Comment:
This struct definition seems to be copied and pasted in 2 places. Can we remove this duplication?

Reply:
True. My thinking is that each `Context` gets its own `testT` struct so that additions to the Context block can be a easily accounted for in `testT`. There is still more duplication of code here than I like, but far less than in the original code, and I want to preserve the "flavor" of the orig code to make reviewing easier. My next pr to address issue 52676 may include some additional refactoring to improve code reuse.